New blog post on retain on delete (#1060)
pulumi/pulumi-hugo: themes/default/content/blog/retainondelete/index.md

# At least one tag is required. Lowercase, hyphen-delimited is recommended.
tags:
    - features

# See the blogging docs at https://github.com/pulumi/docs/blob/master/BLOGGING.md.
# for additional details, and please remove these comments before submitting for review.
---

While Pulumi lets you manage your infrastrucutre with code there are some cases where operations need to
happen outside the standard deployment lifecycle. Pulumi now supports retaining resources on delete allowing a
number of new workflows.

<!--more-->

We've seen a number of user questions over the years about keeping resources after pulumi deployments, or
sharing resources between deployments. All of these worked down to needing a new [resource
option](https://github.com/pulumi/pulumi/issues/7747) to tell the Pulumi engine to not actually delete a
resource when replacing or removing it from the stack.

## Retain on delete

We now support a new resource option [retain on delete]({{< relref "docs/intro/concepts/resources/options/retainOnDelete" >}})
that allows you to specifiy this case to the engine.

When a resource marked with `retainOnDelete` is deleted or replaced by a Pulumi update the engine will not

> call through to the resource providers `Delete` function. This is the similar to AWS Cloud Formation's [Retain

Reviewer: Maybe the engine-providers relationship is too much details here. I would say that it doesn't delete the actual resource in the cloud.

DeletionPolicy](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-attribute-deletionpolicy.html).

> 

Reviewer: Consider adding a link sentence explaining that the following few sections are going to be about scenario examples.

> ## Supporting data storage policies

Reviewer: ```suggestion
### Supporting data storage policies
```

Use cases could be subsections

Using this new feature it is now possible to uncouple resource lifecycles from deployment cycles. This can be

> espically useful to teams with policy around data and storage. For example your company may require that all

Reviewer: ```suggestion
especially useful to teams with policy around data and storage. For example your company may require that all
```

S3 storage is backed up and retained for some amount of time. Using `RetainOnDelete` you can ensure that the
S3 buckets that Pulumi creates aren't deleted by Pulumi, leaving the data alive for your storage policy team
to backup and then delete when no longer required.

## Supporting shared resources

> Another workflow this option enables is the ability to manage externaly owned resources via Pulumi deploments.

Reviewer: ```suggestion
Another workflow this option enables is the ability to manage externally owned resources via Pulumi deployments.
```

Consider running the draft through a spell checker.

You may have an external long lived resource that you want to temporarily manage via a Pulumi program, but
then relquinsh back to the external owner rather than deleting when your finished with it.

Using a combiniation of [import](https://www.pulumi.com/docs/guides/adopting/import/) and `retainOnDelete` you
can bring a resource under managment and then relinqish it back to the external owner.

> ## Working around provider bugs

Reviewer: Should we sugarcoat "bugs"? Sounds a bit harsh :)

Author: problems, issues, failures?

# Draft posts are visible in development, but excluded from production builds.
# Set this property to `false` before submitting your post for review.
draft: true

# Use the meta_desc property to provide a brief summary (one or two sentences)
# of the content of the post, which is useful for targeting search results or social-media
# previews. This field is required or the build will fail the linter test.
meta_desc: Introducing the new resource option retainOnDelete.

# The meta_image appears in social-media previews and on the blog home page.
# A placeholder image representing the recommended format, dimensions and aspect
# ratio has been provided for you.
meta_image: meta.png

# At least one author is required. The values in this list correspond with the `id`
# properties of the team member files at /data/team/team. Create a file for yourself
# if you don't already have one.
authors:
    - fraser-waters

# At least one tag is required. Lowercase, hyphen-delimited is recommended.
tags:
    - features

# See the blogging docs at https://github.com/pulumi/docs/blob/master/BLOGGING.md.
# for additional details, and please remove these comments before submitting for review.
---

> 

Reviewer 2: Slight aside - but reviews on markdown docs in GitHub are quite a bit easier if we don't use hard line breaks for paragraphs.  Then edits can be suggested within a paragraph in a consistent way.  We generally don't use hard line breaks in our hugo site content in part for this reason.

Author: Ah sorry, that's my default for vscode to handle that we do want hard breaks in our code files. I can toggle it off for these posts.

> While Pulumi lets you manage your infrastructure with code there are some cases where operations need to

Reviewer 2: This sounds a little "negative".  I wonder if we can reframe as "Pulumi is frequently used to manage the entire lifecycle of a resource, from creation, to updates, to replacement, to deletion.  However, there are some cases where it is important to ensure that a resource's life can extend *beyond* the lifetime of the Pulumi program that created it.  To support these use cases, Pulumi now support a new resource option `RetainOnDelete` which allows a resource to be retained in the backing cloud provider even after it is delete from the Pulumi stack it is part of."

happen outside the standard deployment lifecycle. Pulumi now supports retaining resources on delete allowing a
number of new workflows.

<!--more-->

We've seen a number of user questions over the years about keeping resources after pulumi deployments, or
sharing resources between deployments. All of these worked down to needing a new [resource
option](https://github.com/pulumi/pulumi/issues/7747) to tell the Pulumi engine to not actually delete a
resource when replacing or removing it from the stack.

## Retain on delete

We now support a new resource option [retain on delete]({{< relref
"docs/intro/concepts/resources/options/retainOnDelete" >}}) that allows you to specify this case to the
engine.

> When a resource marked with `retainOnDelete` is deleted or replaced by a Pulumi update the engine it doesn't

Reviewer 2: ```suggestion
When a resource marked with `retainOnDelete` is deleted or replaced by a Pulumi update, the resource is removed from the Pulumi state but Pulumi does not
```

> delete the actual resource in the cloud. That is it won't call through to the resource providers `Delete`

Reviewer 2: ```suggestion
delete the actual resource in the cloud. That is it won't call through to the resource provider's `Delete`
```

Using this new feature it is now possible to uncouple resource lifecycles from deployment cycles. This can be
especially useful to teams with policy around data and storage. For example your company may require that all
S3 storage is backed up and retained for some amount of time. Using `RetainOnDelete` you can ensure that the
S3 buckets that Pulumi creates aren't deleted by Pulumi, leaving the data alive for your storage policy team
to backup and then delete when no longer required.

### Supporting shared resources

Another workflow this option enables is the ability to manage externally owned resources via Pulumi
deployments. You may have an external long lived resource that you want to temporarily manage via a Pulumi
program, but then relinquish back to the external owner rather than deleting when your finished with it.

Using a combination of [import](https://www.pulumi.com/docs/guides/adopting/import/) and `retainOnDelete` you
can bring a resource under management and then relinquish it back to the external owner.

### Working around provider issues

Lastly not really a new workflow but `retainOnDelete` can also be used to work around some provider issues.

> Some providers will fail to refresh and correctly detect that a resource is now deleted leading pulumi to

Reviewer 2: ```suggestion
Some providers will fail to refresh and correctly detect that a resource is now deleted leading Pulumi to
```

> continually erroring trying to issue a delete call to something that no longer exists. While you could always

Reviewer 2: ```suggestion
error trying to issue a delete call to something that no longer exists. While you could always
```

---
title: "Retain on Delete"
date: 2022-02-23
meta_desc: Introducing the new resource option RetainOnDelete.
meta_image: meta.png
authors:
    - fraser-waters
tags:
    - features

---

> Pulumi is frequently used to manage the entire lifecycle of a resource, from creation, to updates, to replacement, to deletion. However, there are some cases where it is important to ensure that a resource's life can extend beyond the lifetime of the Pulumi program that created it. To support these use cases, Pulumi now support a new resource option `RetainOnDelete` which allows a resource to be retained in the backing cloud provider even after it is delete from the Pulumi stack it is part of.

Reviewer 3: ```suggestion
Pulumi is frequently used to manage the entire lifecycle of a resource, from creation, to updates, to replacement, to deletion. However, there are some cases where it is important to ensure that a resource's life can extend beyond the lifetime of the Pulumi program that created it. To support these use cases, Pulumi now support a new resource option `RetainOnDelete` which allows a resource to be retained in the backing cloud provider even after it is deleted from the Pulumi stack it is part of.
```
